File: desktop/src/main/kotlin/com/videoroom/viewmodel/DetailViewModel.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 VideoRoom Contributors

package com.videoroom.viewmodel

import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import com.videoroom.data.models.VideoMetadata
import com.videoroom.data.repository.VideoRepository
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.slf4j.LoggerFactory

class DetailViewModel(
    private val repository: VideoRepository
) {
    private val logger = LoggerFactory.getLogger(DetailViewModel::class.java)
    private val viewModelScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val _metadata = mutableStateOf<VideoMetadata?>(null)
    val metadata: State<VideoMetadata?> = _metadata

    private val _thumbnail = mutableStateOf<ByteArray?>(null)
    val thumbnail: State<ByteArray?> = _thumbnail

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _notes = MutableStateFlow("")
    val notes: StateFlow<String> = _notes.asStateFlow()

    // Group / stack info
    private val _groupMembers = MutableStateFlow<List<com.videoroom.data.models.VideoSummary>>(emptyList())
    val groupMembers: StateFlow<List<com.videoroom.data.models.VideoSummary>> = _groupMembers.asStateFlow()

    private val _groupPreferredId = MutableStateFlow("")
    val groupPreferredId: StateFlow<String> = _groupPreferredId.asStateFlow()

    private var currentVideoSummary: com.videoroom.data.models.VideoSummary? = null

    fun setCurrentVideo(video: com.videoroom.data.models.VideoSummary) {
        currentVideoSummary = video
        if (video.isInGroup) {
            loadGroupMembers(video.groupId)
        } else {
            _groupMembers.value = emptyList()
            _groupPreferredId.value = ""
        }
    }

    private fun loadGroupMembers(groupId: String) {
        viewModelScope.launch {
            try {
                val (members, preferred) = repository.listGroupMembers(groupId)
                _groupMembers.value = members
                _groupPreferredId.value = preferred
            } catch (e: Exception) {
                logger.warn("Failed to load group members for $groupId", e)
            }
        }
    }

    fun setGroupPreferred(videoId: String) {
        val groupId = currentVideoSummary?.groupId ?: return
        if (groupId.isEmpty()) return
        viewModelScope.launch {
            try {
                if (repository.setGroupPreferred(groupId, videoId)) {
                    _groupPreferredId.value = videoId
                    logger.info("Set preferred video to $videoId")
                } else {
                    _error.value = "Failed to set preferred video"
                }
            } catch (e: Exception) {
                _error.value = "Failed to set preferred: ${e.message}"
            }
        }
    }

    /** Remove the currently-displayed video from its stack. `onComplete`
     *  fires with the *old* group id after the daemon call succeeds —
     *  callers wire it to `GridViewModel.refreshAfterStackChange` so the
     *  grid's representative + expanded-member caches catch up. */
    fun ungroupCurrent(onComplete: (oldGroupId: String) -> Unit = {}) {
        val videoId = _metadata.value?.id ?: return
        // Capture the *current* group id before we tell the daemon to
        // drop it — VideoMetadata doesn't carry groupId, so we read it
        // off the VideoSummary we cached on selection.
        val oldGroupId = currentVideoSummary?.groupId ?: ""
        viewModelScope.launch {
            try {
                if (repository.ungroupVideo(videoId)) {
                    _groupMembers.value = emptyList()
                    _groupPreferredId.value = ""
                    logger.info("Ungrouped video $videoId")
                    onComplete(oldGroupId)
                } else {
                    _error.value = "Failed to ungroup"
                }
            } catch (e: Exception) {
                _error.value = "Failed to ungroup: ${e.message}"
            }
        }
    }

    fun loadMetadata(videoId: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            try {
                val metadata = repository.getVideoMetadata(videoId)
                _metadata.value = metadata
                _notes.value = metadata?.notes ?: ""

                if (metadata != null) {
                    // Load thumbnail
                    loadThumbnail(videoId)
                    logger.info("Loaded metadata for video: ${metadata.filename}")
                } else {
                    _error.value = "Video not found"
                }
                _isLoading.value = false
            } catch (e: Exception) {
                _error.value = "Failed to load metadata: ${e.message}"
                _isLoading.value = false
                logger.error("Failed to load metadata for video: $videoId", e)
            }
        }
    }

    private fun loadThumbnail(videoId: String) {
        viewModelScope.launch {
            try {
                val thumbnail = repository.getThumbnail(videoId, "large")
                _thumbnail.value = thumbnail
                if (thumbnail != null) {
                    logger.info("Loaded thumbnail for video: $videoId")
                }
            } catch (e: Exception) {
                logger.warn("Failed to load thumbnail for video: $videoId", e)
                // Don't error out, just skip thumbnail
            }
        }
    }

    fun updateNotes(newNotes: String) {
        _notes.value = newNotes
        val videoId = _metadata.value?.id ?: return

        viewModelScope.launch {
            try {
                val success = repository.updateVideoNotes(videoId, newNotes)
                if (!success) {
                    _error.value = "Failed to update notes"
                    logger.warn("Failed to update notes for video: $videoId")
                } else {
                    logger.info("Updated notes for video: $videoId")
                }
            } catch (e: Exception) {
                _error.value = "Failed to update notes: ${e.message}"
                logger.error("Failed to update notes for video: $videoId", e)
            }
        }
    }

    fun addTag(tagId: String) {
        val videoId = _metadata.value?.id ?: return

        viewModelScope.launch {
            try {
                val success = repository.tagVideos(listOf(videoId), tagId)
                if (success) {
                    // Reload metadata to get updated tags
                    loadMetadata(videoId)
                    logger.info("Added tag to video: $videoId")
                } else {
                    _error.value = "Failed to add tag"
                }
            } catch (e: Exception) {
                _error.value = "Failed to add tag: ${e.message}"
                logger.error("Failed to add tag to video: $videoId", e)
            }
        }
    }

    fun removeTag(tagId: String) {
        val videoId = _metadata.value?.id ?: return

        viewModelScope.launch {
            try {
                val success = repository.untagVideos(listOf(videoId), tagId)
                if (success) {
                    // Reload metadata to get updated tags
                    loadMetadata(videoId)
                    logger.info("Removed tag from video: $videoId")
                } else {
                    _error.value = "Failed to remove tag"
                }
            } catch (e: Exception) {
                _error.value = "Failed to remove tag: ${e.message}"
                logger.error("Failed to remove tag from video: $videoId", e)
            }
        }
    }

    fun addToCollection(collectionId: String) {
        val videoId = _metadata.value?.id ?: return

        viewModelScope.launch {
            try {
                val success = repository.addToCollection(listOf(videoId), collectionId)
                if (success) {
                    // Reload metadata to get updated collections
                    loadMetadata(videoId)
                    logger.info("Added video to collection: $collectionId")
                } else {
                    _error.value = "Failed to add to collection"
                }
            } catch (e: Exception) {
                _error.value = "Failed to add to collection: ${e.message}"
                logger.error("Failed to add video to collection: $collectionId", e)
            }
        }
    }

    fun clearError() {
        _error.value = null
    }

    fun clear() {
        _metadata.value = null
        _thumbnail.value = null
        _notes.value = ""
        _error.value = null
    }

    fun onDestroy() {
        viewModelScope.cancel()
        logger.info("DetailViewModel destroyed")
    }
}

File: macos/ReelVault/ViewModels/DetailViewModel.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

import SwiftUI
import Combine

@MainActor
class DetailViewModel: ObservableObject {
    @Published var metadata: VideoMetadata?
    @Published var thumbnail: NSImage?
    @Published var isLoading = false
    @Published var error: String?
    @Published var notes = ""

    @Published var groupMembers: [VideoSummary] = []
    @Published var groupPreferredId: String = ""

    /// Proxies attached to the currently-selected video. Sorted by
    /// pixel count descending (server contract). UI sites can read
    /// `.last` for the smallest proxy (used by inline grid/list
    /// playback) or `.first` for the largest.
    @Published var proxies: [VideoRepository.ProxyInfo] = []

    /// ID of the proxy the user explicitly picked for full-screen
    /// detail-view playback. `nil` means "play the master" (with the
    /// implicit unplayable-master → smallest-proxy fallback applied by
    /// `playbackPath(for:)`).
    @Published var selectedProxyId: String?

    /// Content for the top-bar proxy-playback indicator.
    struct ProxyBanner: Equatable {
        /// True when the user explicitly picked this proxy in the right
        /// panel; false when it's the automatic unplayable-master fallback.
        var selected: Bool
        /// "filename • 1080p"-style detail, or nil when the proxy row is
        /// unknown. Surfaced in the indicator's tooltip.
        var detail: String?
    }

    /// Drives the proxy-playback indicator that now lives in the top bar
    /// (see `ContentView.topBar`) rather than as an overlay on the video.
    /// `nil` hides it — the master is playing, or the loupe isn't on screen.
    /// `DetailLoupeView` sets this as it resolves the effective playback path
    /// and clears it when the loupe is left.
    @Published var proxyBanner: ProxyBanner?

    /// The proxy the loupe is actually playing right now (auto-chosen or the
    /// user's pick), so the right-panel list can highlight which one is playing
    /// by default — without pinning `selectedProxyId` (which stays the user's
    /// explicit choice so they can still revert to the master).
    @Published var playingProxyId: String?

    /// The summary backing the currently-selected card. Exposed so the
    /// right panel's proxy section can decide whether to render itself
    /// (keys off `hasProxies` and `playableNatively`, neither of which
    /// is carried by `VideoMetadata`).
    @Published var currentSummary: VideoSummary?

    private var currentVideoSummary: VideoSummary? { currentSummary }
    private let repository = VideoRepository.shared

    /// Called by the grid when the user selects a video — drives the group/stack
    /// section in addition to the regular metadata load.
    func setCurrentVideo(_ video: VideoSummary) {
        currentSummary = video
        if video.isInGroup {
            loadGroupMembers(groupId: video.groupId)
        } else {
            groupMembers = []
            groupPreferredId = ""
        }
        // Reset proxy selection on each new video, then lazily load the
        // proxy list when the summary advertises any. Servers send
        // `proxyCount == 0` for most rows, so the conditional saves a
        // round trip on the common case.
        selectedProxyId = nil
        if video.hasProxies {
            loadProxies(videoId: video.id)
        } else {
            proxies = []
        }
    }

    private func loadProxies(videoId: String) {
        Task {
            do {
                proxies = try await repository.listProxies(videoId: videoId)
            } catch {
                NSLog("Failed to load proxies for \(videoId): \(error)")
                proxies = []
            }
        }
    }

    /// User picked a specific proxy in the detail-view right panel.
    /// Pass `nil` to revert to playing the master.
    func setSelectedProxy(_ proxyId: String?) {
        selectedProxyId = proxyId
    }

    /// Recompute the top-bar proxy indicator from the current proxy
    /// selection and the player's render height. Called by `DetailLoupeView`
    /// whenever the effective path could change (selection, proxy list,
    /// render-area resize). Sets `proxyBanner` to nil when the master plays.
    func updateProxyBanner(for video: VideoSummary, areaHeightPx: Int) {
        let effective = playbackPath(for: video, areaHeightPx: areaHeightPx) ?? video.path
        if effective != video.path {
            let active = proxies.first(where: { $0.path == effective })
            proxyBanner = ProxyBanner(
                selected: selectedProxyId != nil,
                detail: active.map { "\($0.filename) • \($0.height)p" }
            )
            playingProxyId = active?.id
        } else {
            proxyBanner = nil
            playingProxyId = nil
        }
    }

    /// Hide the top-bar proxy indicator (loupe left, or selection cleared).
    func clearProxyBanner() {
        proxyBanner = nil
        playingProxyId = nil
    }

    /// Break the link between the currently-displayed master and one
    /// of its proxies. Refreshes the proxy list on success.
    func breakProxyLink(proxyId: String, onChanged: @escaping () -> Void = {}) {
        guard let masterId = currentSummary?.id else { return }
        Task {
            let ok = await repository.removeProxyLink(masterId: masterId, proxyId: proxyId)
            if ok {
                do {
                    proxies = try await repository.listProxies(videoId: masterId)
                } catch {
                    NSLog("Reload proxies after break failed: \(error)")
                }
                if selectedProxyId == proxyId { selectedProxyId = nil }
                // A detached proxy becomes a standalone video again — let the
                // grid + left-panel counts catch up.
                onChanged()
            } else {
                error = "Failed to remove proxy link"
            }
        }
    }

    /// Mark `proxyId` as a manual proxy of the currently-displayed
    /// master. Surfaces via the inspector's "Add selected video as
    /// proxy" affordance.
    func forceProxyLink(proxyId: String, onChanged: @escaping () -> Void = {}) {
        guard let masterId = currentSummary?.id else { return }
        guard masterId != proxyId else {
            error = "A video can't be a proxy of itself"
            return
        }
        Task {
            let ok = await repository.setProxyOf(proxyId: proxyId, originalId: masterId)
            if ok {
                do {
                    proxies = try await repository.listProxies(videoId: masterId)
                } catch {
                    NSLog("Reload proxies after force failed: \(error)")
                }
                // A newly-attached proxy stops counting as a standalone video —
                // refresh the grid + left-panel counts.
                onChanged()
            } else {
                error = "Failed to add proxy link"
            }
        }
    }

    /// The on-disk path the detail-view player should load, based on the
    /// current proxy selection and the player's render height `areaHeightPx`
    /// (in px):
    ///   * User explicitly picked a proxy → that proxy's path.
    ///   * A proxy exists → the proxy whose resolution best matches the render
    ///     area (see `chooseProxy(areaHeightPx:)`). We prefer a proxy even for
    ///     a natively-playable master — detail playback defaults to a proxy,
    ///     with the master one click away via the right-panel picker.
    ///   * Otherwise → nil; the caller falls back to the master path.
    func playbackPath(for video: VideoSummary, areaHeightPx: Int = 0) -> String? {
        if let picked = selectedProxyId,
           let row = proxies.first(where: { $0.id == picked }) {
            return row.path
        }
        guard !proxies.isEmpty else { return nil }
        // When the master itself can't play locally (above the configured
        // playable height), restrict to proxies that *can* play — so detail
        // playback defaults to something that actually decodes. If none
        // qualifies, fall back to all (chooseProxy's smallest).
        var candidates = proxies
        if !video.playableNatively {
            let playable = proxies.filter { $0.playableNatively }
            if !playable.isEmpty { candidates = playable }
        }
        return chooseProxy(from: candidates, areaHeightPx: areaHeightPx)?.path
    }

    /// Pick the proxy that best fills a render area `areaHeightPx` px tall: the
    /// smallest proxy still at least as tall as the area (so it isn't upscaled),
    /// or — when the area is taller than every proxy — the largest. Before the
    /// area is measured (`areaHeightPx <= 0`) we fall back to the smallest.
    /// Returns nil when `candidates` is empty. `proxies` is sorted descending by
    /// pixel count (first = largest, last = smallest); `candidates` preserves
    /// that order.
    private func chooseProxy(
        from candidates: [VideoRepository.ProxyInfo],
        areaHeightPx: Int
    ) -> VideoRepository.ProxyInfo? {
        guard !candidates.isEmpty else { return nil }
        if areaHeightPx <= 0 { return candidates.last }
        return candidates.filter { $0.height >= areaHeightPx }.min(by: { $0.height < $1.height })
            ?? candidates.first
    }

    private func loadGroupMembers(groupId: String) {
        Task {
            do {
                let (members, preferred) = try await repository.listGroupMembers(groupId: groupId)
                groupMembers = members
                groupPreferredId = preferred
            } catch {
                NSLog("Failed to load group members: \(error)")
            }
        }
    }

    func setGroupPreferred(videoId: String) {
        guard let groupId = currentVideoSummary?.groupId, !groupId.isEmpty else { return }
        Task {
            do {
                if try await repository.setGroupPreferred(groupId: groupId, videoId: videoId) {
                    groupPreferredId = videoId
                } else {
                    error = "Failed to set preferred video"
                }
            } catch {
                self.error = "Failed to set preferred: \(error.localizedDescription)"
            }
        }
    }

    /// Remove the currently-displayed video from its stack. `onComplete`
    /// fires after the daemon call succeeds — callers wire it to
    /// `GridViewModel.refreshAfterStackChange(groupId:)` so the grid's
    /// representative + expanded-member caches catch up.
    func ungroupCurrent(onComplete: ((_ groupId: String) -> Void)? = nil) {
        guard let videoId = metadata?.id else { return }
        // Capture the *current* group id before we tell the daemon to
        // drop it — `VideoMetadata` doesn't track groupId, so we read it
        // from the `VideoSummary` we cached on selection.
        let oldGroupId = currentVideoSummary?.groupId ?? ""
        Task {
            do {
                if try await repository.ungroupVideo(videoId: videoId) {
                    groupMembers = []
                    groupPreferredId = ""
                    onComplete?(oldGroupId)
                } else {
                    error = "Failed to ungroup"
                }
            } catch {
                self.error = "Failed to ungroup: \(error.localizedDescription)"
            }
        }
    }

    func loadMetadata(videoId: String) {
        isLoading = true
        error = nil

        Task {
            do {
                let meta = try await repository.getVideoMetadata(videoId: videoId)
                self.metadata = meta
                self.notes = meta.notes
                self.isLoading = false
                await loadThumbnail(videoId: videoId)
            } catch {
                self.error = "Failed to load metadata: \(error.localizedDescription)"
                self.isLoading = false
            }
        }
    }

    private func loadThumbnail(videoId: String) async {
        do {
            thumbnail = try await repository.getThumbnail(videoId: videoId, size: "large")
        } catch {
            NSLog("Failed to load thumbnail: \(error)")
        }
    }

    func updateNotes(_ newNotes: String) {
        notes = newNotes
        guard let videoId = metadata?.id else { return }
        Task {
            do {
                let ok = try await repository.updateVideoNotes(videoId: videoId, notes: newNotes)
                if !ok { error = "Failed to update notes" }
            } catch {
                self.error = "Failed to update notes: \(error.localizedDescription)"
            }
        }
    }

    func clearError() { error = nil }

    func clear() {
        metadata = nil
        thumbnail = nil
        notes = ""
        error = nil
        groupMembers = []
        groupPreferredId = ""
        currentSummary = nil
        proxies = []
        selectedProxyId = nil
    }
}
